[ASCII-1549] improve the overview and creating a component page (#25028)
DataDog/datadog-agent: docs/public/components/creating-components.md

# Creating a Component

This page explains how to create components in detail.

> Thought this page we're going to create a compression component. The component is going to handle the payload compression before it gets sent to the Datadog backend.

Reviewer: ```suggestion
This page uses the process of creating a compression component as an example. The component compresses the payload before sending it to the Datadog backend.
```

> Since there are multiple ways to compress data we are going to provide two implementations of the same component interface. 

Reviewer: ```suggestion
Since there are multiple ways to compress data, we provide two implementations of the same component interface: 
```

> * one using [ZSTD](https://en.wikipedia.org/wiki/Zstd) data compression algorithm.
> * one using [ZIP](https://en.wikipedia.org/wiki/ZIP_(file_format)) data compression algorithm.

Reviewer: ```suggestion
* the [ZSTD](https://en.wikipedia.org/wiki/Zstd) data compression algorithm
* the [ZIP](https://en.wikipedia.org/wiki/ZIP_(file_format)) data compression algorithm
```

A Component is organize into multiple folders and Go packages. The main reason to separate a component into separate package is to separate the interface from the implementations and to improve code sharing. Having the interface declare in a separate package from the implementation allow to import the interface without having to import the different implementations. 

## File hierarchy

All components are located in the `comp` folder at the top of the Agent repo.

The file hierarchy is the following:

```
comp /

>   <bundle name> /        <-- Optional

Author: Yes they are https://github.com/DataDog/datadog-agent/commit/a0d372e985934f37f269c997e04441d5f6e35a0b 😉

      impl-<alternate> / <-- An alternate implementation.
      impl-none /        <-- Optional. A none implementation.
      fx /               <-- All fx related logic for the primary implementation, if any.
      fx-alternate /     <-- All fx related logic for a specific implementation.
      mock /             <-- The mock of the component to ease testing.
      fx-<alternate> /   <-- All fx related logic for a specific implementation.
      mock /             <-- The mock interface of the component to ease testing.
```

TODO: where do we put the `optional.NoneOption` ? Into its own fx-none folder ?

To note:

* If your component has only one implementation it should live in the `impl` folder.
* If your component don't have a primary implementation but several version you should have no `impl` folder but
* If your component don't have a single implementation but several versions you should have no `impl` folder but
  multiple `impl-<version>` folders. For example a component in charge of compressing data will have `impl-zstd` and `impl-zip` but not `impl` folder.
* If your component needs to offer a dummy/empty version it should live in the `impl-none` folder.

> * A mock version is mandatory for any component exposing at least one method.

Reviewer 2: I think we should keep the mock line

* A main that imports a component should be able to select a specific implementation without compiling with the others.
  For example: the ZSTD library should not be included at compile time when the ZIP version is used.

!!! warning "Important"

>     For all these reasons, components should never be nested.

Reviewer 2: Why not ?

All implementations must use the package name: `<component name>impl`.

For example, a compression component with 2 implementation would use `package compressionimpl` in both
`comp/<bundle>/compression/impl-zstd` and `comp/<bundle>/compression/impl-zip` folders.

> 

Reviewer 2: Actually, after trying to migrate a package, this is so annoying as it force import aliasing everywhere and we only import implementation `fx` folder and main. We should remove this

> ## Bootstraping Components

Reviewer: ```suggestion
## Bootstrapping components
```

For example, a compression component with 2 implementation would use `package compressionimpl` in both
`comp/<bundle>/compression/impl-zstd` and `comp/<bundle>/compression/impl-zip` folders.

## Bootstraping Components

You can use the [invoke](../setup.md#invoke) task `inv components.new-component comp/<component>` to generate a scaffold for your new component.

Every public variable, function, struct, and interface of your component **must** be documented. Please refer to the [Documentation](#documentation) section below for details.

### The def folder

The def folder will contain your interface and ALL public types needed by the users of your component.

For our compression example is will look like this:
For our compression example it will look like this:

> === ":octicons-file-code-16: comp/&lt;bundleName&gt;/compression/def/component.go"

Author: I removed the `&lt;bundleName&gt;` to avoid adding cognitive overload to the readers. Since at this stage of the documentation they do not know about bundles

Reviewer 2: true, but I think bundle are mandatory and people might copy past the the path,

    }
    ```

> All component interfaces must be called `Component` so all imports would be similar to `compressiondef.Component`.

Reviewer: ```suggestion
All component interfaces must be called `Component`, so all imports have the form `compressiondef.Component`.
```

You can see that our interface only exposes the bare minimum. You should aim at having the smallest possible interface
for your component. 

> When defining a component interface we try to use struct or interface from third-party dependencies. Ensuring that using the component interface doesn't require importing any thrid-party dependency.

Reviewer: ```suggestion
When defining a component interface, use structs or interfaces from third-party dependencies. Using structs or interfaces allows you to use the component interface without importing a third-party dependency.
```

!!! warning "Interface using a third-party dependency"
    ```
    package telemetrydef

    import "github.com/prometheus/client_golang/prometheus"

    // team: agent-shared-components

    // Component is the component type.
    type Component interface {
        // RegisterCollector Registers a Collector with the prometheus registry
        RegisterCollector(c prometheus.Collector)
    }
    ```

> For the example above every user of the telemetry component would have to import `github.com/prometheus/client_golang/prometheus`. To avoid those cases is best to design your interface mindfully.
> 
> TODO: write alternative solution

Reviewer 2: ```suggestion
For the example above every user of the `telemetry` component would have to import `github.com/prometheus/client_golang/prometheus` no matter which implementation they use.

In general, be mindful of using external types in the public interface of your component. For example, it would make sens to use docker types in a `docker` component but not in a `container` component.
```

> Also note that there is no `Start`/`Stop` method. Anything related to lifecycle will be handle
> internally by each component (more on this [here TODO]()).

Reviewer: ```suggestion
Also note that there is no `Start` or `Stop` method. Each component handles its lifecycle internally. (more on this [here TODO]()).
```

TODO: write the lifecycle part and update the link above

### The impl folders

> The `impl` folder is where the component implementation is written. The component implementation details is up to the developer to design. 

Reviewer 2: ```suggestion
The `impl` folder is where the component implementation is written. The component implementation details is up to the developer. 
```

The only requirement is that there is a public instantiation function called `NewComponent`.

=== ":octicons-file-code-16: comp/compression/fx-zstd/component.go"
    ```go
    package zstdimpl

    import (
        // We always import the component def folder to be able to return a 'def.Component' type.
        "github.com/DataDog/datadog-agent/comp/compression/def"
    )

    // NewComponent returns a new ZSTD implementation for the compression component
    func NewComponent(){
        ....
    }
    ```

To access an argument in the  `NewComponent` function we use a special struct name `Requires`. And the return value is a special stuct named `Provides`. Those are internal nomenclature used to handle the different component depencies using `FX`

> For our component we are going to need to access the configuration component and the log component. To express that in our component we are going to define a `Requires` struct with two fields. The name is irrelevant but the type must by a concrete type. 

Reviewer 2: This seems more convoluted than the original one liner, no ? We could add a line saying that FX/dependencies injection is optional but not `Requires` and `Provides` ?

        Conf config.Component
        Log  log.Component
    }
    ```
    
!!! Info "Using other components"
    You want to use another component within your own? Simply add it to the `Requires` struct and `Fx` will give it to
    you at initialization ! Be careful of cycling dependencies though.
    

> As for the output of our component we are going populate the `Provides` struct with the all the return values.

Reviewer 2: I think 1 code example is better than 3. We should explain in 1 paragraph the `Provides` and `Requires` and then show a single example of code. Most people will only look at the code snippet and it will have everything they need.
